staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go

```diff
@@ -0,0 +1,379 @@
+}
+
+// FlowDistinguisherMethodType is the type of flow distinguisher method
+type FlowDistinguisherMethodType string
+
+// FlowDistinguisherMethod specifies the method of a flow distinguisher.
+type FlowDistinguisherMethod struct {
+	// `type` is the type of flow distinguisher method
+	// The supported types are "ByUser" and "ByNamespace".
+	// Required.
+	Type FlowDistinguisherMethodType `json:"type" protobuf:"bytes,1,opt,name=type"`
+}
+
+// PriorityLevelConfigurationReference contains information that points to the "request-priority" being used.
+type PriorityLevelConfigurationReference struct {
+	// `name` is the name of resource being referenced
+	// Required.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+}
+
+// PolicyRulesWithSubjects prescribes a test that applies to a request to an apiserver. The test considers the subject
+// making the request, the verb being requested, and the resource to be acted upon. This PolicyRulesWithSubjects matches
+// a request if and only if both (a) at least one member of subjects matches the request and (b) at least one member
+// of resourceRules or nonResourceRules matches the request.
+type PolicyRulesWithSubjects struct {
+	// `subjects` is the list of normal user, serviceaccount, or group that this rule cares about.
+	// There must be at least one subject for this field, or it will be rejected by the api validation.
+	// Required.
+	Subjects []Subject `json:"subjects" protobuf:"bytes,1,rep,name=subjects"`
+	// `resourceRules` is a slice of ResourcePolicyRules that identify matching requests according to their verb and the
+	// resource they request to act upon.
+	// +optional
+	ResourceRules []ResourcePolicyRule `json:"resourceRules,omitempty" protobuf:"bytes,2,opt,name=resourceRules"`
+	// `nonResourceRules` is a list of NonResourcePolicyRules that identify matching requests according to their verb
+	// and the non-resource URL they request to act upon.
+	// +optional
+	NonResourceRules []NonResourcePolicyRule `json:"nonResourceRules,omitempty" protobuf:"bytes,3,opt,name=nonResourceRules"`
+}
+
+// Subject matches a set of users.
+// Syntactically, Subject is a general API object reference.
+// Authorization produces a username and a set of groups, and we imagine special kinds of non-namespaced objects,
+// User and Group to represent such a username or group.
+// The only kind of true object reference that currently will match any users is ServiceAccount.
+type Subject struct {
+	// `kind` of object being referenced. Values defined by this API group are "User", "Group", and "ServiceAccount".
+	// If the kind value is not recognized, the flow-control layer in api-server should report an error.
+	// Required.
+	Kind string `json:"kind" protobuf:"bytes,1,opt,name=kind"`
+	// `name` of the object being referenced.  To match regardless of name, use '*'.
+	// Required.
+	Name string `json:"name" protobuf:"bytes,2,opt,name=name"`
```

Comment:
in a true union type, this and namespace would be grouped by stanza and distinct.

Reply:
Are you suggesting something like the following?
```
// +union
type Subject struct {
    // Required
    // +unionDiscriminator
    Kind SubjectKind `json:"kind" protobuf:"bytes,1,opt,name=kind"`
    // +optional
    User *NamedSubject `json:"user,omitempty" protobuf:"bytes,2,opt,name=user"`
    // +optional
    Group *NamedSubject `json:"group,omitempty" protobuf:"bytes,3,opt,name=group"`
    // +optional
    ServiceAccount *NamespaceNamedSubject `json:"serviceAccount,omitempty" protobuf:"bytes,4,opt,name=serviceAccount"`
}

type SubjectKind string
const SubjectKindUser SubjectKind = "User"
const SubjectKindGroup SubjectKind = "Group"
const SubjectKindServiceAccount SubjectKind = "ServiceAccount"

type NamedSubject struct {
    // Required
    Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
}

type NamespaceNamedSubject struct {
    // Required
    Namespace string `json:"namespace" protobuf:"bytes,1,opt,name=namespace"`
    // Required
    Name string `json:"name" protobuf:"bytes,2,opt,name=name"`
}
```

Reply:
i updated the subject to union type.